### PR TITLE
Add WinClosed event

### PR DESCRIPTION
**Pull request by obcat**
- Ported form Neovim:
  - https://github.com/neovim/neovim/pull/11331
  - https://github.com/neovim/neovim/commit/acaf5d233b9128c4b494cec6ca1bf090dc2f7a56

- Maybe #742 can be closed by this patch.

### CODE REVIEW

**Comment by obcat** [comment]
Oh, AppVeyor build is failed... I'll take a look tommorow.

**Comment by lacygoill** [comment]
> Maybe Autocommand event for when a window is closed #742 can be closed by this patch.

This todo item might also be closed if the PR gets merged (line 4882 in `runtime/doc/todo.txt`):

    WinClose	    - Just before closing a window

[source](https://github.com/vim/vim/blob/113cb513f76d8866cbb6dc85fa18aded753e01da/runtime/doc/todo.txt#L4882)

**Comment by obcat** [comment]
> This todo item might also be closed if the PR gets merged (line 4882 in `runtime/doc/todo.txt`):
> 
> ```
> WinClose	    - Just before closing a window
> ```
> 
> [source](https://github.com/vim/vim/blob/113cb513f76d8866cbb6dc85fa18aded753e01da/runtime/doc/todo.txt#L4882)

That says "**before** closing a window", but the `WinClosed` event is made to be fired "**after** closing a window".
But yes, if `WinClosed` is enough, that todo item might alse be closed.

File: src/window.c [hunk 1]
```diff
@@ -2566,6 +2567,13 @@ win_close(win_T *win, int free_buf)
     if (popup_win_closed(win) && !win_valid(win))
 	return FAIL;
 #endif
+
+    // Trigger WinClosed just before starting to free window-related resources.
+    trigger_winclosed(win);
+    // autocmd may have freed the window already.
+    if (!win_valid_any_tab(win))
+	return OK;
```
Comment (obcat) [inline src/window.c:2575]: I don't know how to cover this line by tests...

File: src/window.c [hunk 2]
```diff
@@ -2731,6 +2753,12 @@ win_close_othertab(win_T *win, int free_buf, tabpage_T *tp)
 					       && win->w_buffer->b_locked > 0))
 	return; // window is already being closed
 
+    // Trigger WinClosed just before starting to free window-related resources.
+    trigger_winclosed(win);
+    // autocmd may have freed the window already.
+    if (!win_valid_any_tab(win))
+	return;
```
Comment (obcat) [inline src/window.c:2760]: ditto

**Comment by obcat** [comment]
Very useful! Thank you!

**Pull request closed without merging**